File: bileton-trip/src/main/java/com/ademozalp/bileton/model/Trip.java
package com.ademozalp.bileton.model;

import com.ademozalp.bileton.model.base.BaseEntity;
import com.ademozalp.bileton.util.DBConstant;
import com.ademozalp.bu_bilet.model.enums.TripStatus;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Formula;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.Duration;
import java.time.LocalDateTime;

import static com.ademozalp.bileton.util.DBConstant.*;

@Entity
@Table(name = DBConstant.TRIP)
@EntityListeners(AuditingEntityListener.class)
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Builder
public class Trip extends BaseEntity {
    @ManyToOne
    @JoinColumn(name = ORIGIN, nullable = false)
    private Location origin;
    @ManyToOne
    @JoinColumn(name = DESTINATION, nullable = false)
    private Location destination;
    @Column(name = DEPARTURE_TIME, nullable = false)
    private LocalDateTime departureTime;
    @Column(name = ARRIVAL_TIME, nullable = false)
    private LocalDateTime arrivalTime;
    @Formula("EXTRACT(EPOCH FROM (arrival_time - departure_time))")
    private Long durationInSeconds;
    @Builder.Default
    @Enumerated(EnumType.STRING)
    @Column(name = TRIP_STATUS, nullable = false)
    private TripStatus tripStatus = TripStatus.PENDING;

    @ManyToOne
    @JoinColumn(name = DBConstant.COMPANY_ID, nullable = false)
    private Company company;

    public Duration getDuration() {
        return Duration.ofSeconds(durationInSeconds);
    }
}

File: bileton-trip/src/main/java/com/ademozalp/bileton/model/Trip.kt
package com.ademozalp.bileton.model

import com.ademozalp.bileton.util.DatabaseConstant
import com.ademozalp.bu_bilet.model.enums.TripStatusType
import com.ademozalp.bu_bilet.model.enums.VehicleType
import jakarta.persistence.*
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.math.BigDecimal
import java.time.LocalDateTime


@Entity
@Table(name = DatabaseConstant.TRIP)
@EntityListeners(AuditingEntityListener::class)
data class Trip @JvmOverloads constructor(

    @Column(nullable = false)
    var price: BigDecimal,
    @Column(nullable = false)
    var duration: String,

    @Enumerated(EnumType.STRING)
    var status: TripStatusType? = TripStatusType.ACTIVE,

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    var vehicleType: VehicleType,

    @Column(nullable = false)
    var departureDate: LocalDateTime,

//    @ManyToOne
//    @JoinColumn(name = DatabaseConstant.DEPARTURE_CITY_ID, nullable = false)
//    var departureCity: City,
//
//    @ManyToOne
//    @JoinColumn(name = DatabaseConstant.DESTINATION_CITY_ID, nullable = false)
//    var destinationCity: City,
//
//    @ManyToOne
//    @JoinColumn(name = DatabaseConstant.COMPANY_ID, nullable = false)
//    var company: Company,
//
//
//    @OneToMany(mappedBy = "trip", cascade = [CascadeType.REMOVE], fetch = FetchType.EAGER)
//    var tickets: List<Ticket> = mutableListOf()


) : BaseEntity() {
    constructor() : this(
        BigDecimal.ZERO,
        "0:0",
        TripStatusType.ACTIVE,
        VehicleType.BUS,
        LocalDateTime.now(),
//        City(),
//        City(),
//        Company()
    )


    override fun hashCode(): Int {
        return super.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        return super.equals(other)
    }
}
